3.6 - Add a deprecation warning helper.

markstory opened the pull request:
This will make adding framework deprecation warnings we plan to add in 3.6.x simpler as they will be easier to write.

markstory committed:
Add a deprecation warning helper.

This will make adding framework deprecation warnings we plan to add in
3.6.x simpler as they will be easier to write.

chinpei215 commented:
I think the function would report unhelpful line number and file name.

inoas commented:
```php
<?php

    function deprecationWarning($message)
    {
        trigger_error($message, E_USER_DEPRECATED);
    }

deprecationWarning('here');
```

PHP Deprecated:  here in /Users/leo/Downloads/test-error.php on line 5

Deprecated: here in /Users/leo/Downloads/test-error.php on line 5

markstory commented:
@chinpei215 The HTML error output will include a backtrace. Would you want to see the backtrace in the deprecation warning message as well?

dereuromark commented:
the log entries would be useful to have the right location and line number.

markstory commented:
@dereuromark Logging is something that the error handler already does if its configured correctly.

markstory committed:
Fix failing test.

Test was failing because error_reporting level was modified by another
test.

chinpei215 commented:
@markstory But people would have to click links to show stacktraces when they encountered errors.
![deprecations](https://user-images.githubusercontent.com/7399393/29598847-a4b617ac-8806-11e7-8f1a-db50b937fc0c.png)

markstory commented:
@chinpei215 Fair, but they would have to do that anyways as the raw `trigger_error` calls would point to code inside cakephp not their application. We could do what @dereuromark suggested and 'jump up' two stack frames to hopefully get the right line.

chinpei215 commented:
@markstory ah sorry you are right. and it might be nice if we could do that.

markstory commented:
I think generating a backtrace 2 levels up is reasonable. However, I would want to guard that kind of logic with a debug check. Does that seem reasonable?

chinpei215 commented:
Sounds reasonable to me.

markstory committed:
Add stackframe information to deprecation message.

Having the userland method line/file should help in fixing deprecation
warnings. I've not used Debugger here as it would add a dependency to
cakephp/core that will not be easy to satisfy for standalone packages.

markstory commented:
Error messages now have file/lines displayed.

gmponos commented:
IMHO I see more trouble caused rather than value added.

markstory commented:
@gmponos If we use trigger_error() without a wrapper, the errors will not be able to indicate what application code is using the deprecated functions.

gmponos commented:
> If we use trigger_error() without a wrapper, the errors will not be able to indicate what application code is using the deprecated functions.

If that's the goal then fine. I maybe rushed judging the goal of this PR because of the initial commit where the only thing I saw was a "proxy" function...

But still your comment raises some more questions for me and I am sorry for that... maybe because I don't follow all discussions.

I believe you are using an ErrorHandler on a CakePHP application isn't this possible to be handled there? Hope I didn't misunderstood and sorry for any trouble.

markstory commented:
> I believe you are using an ErrorHandler/Debugger on a CakePHP application isn't this possible to be handled there?

We are using an error handler, but that error handler relies on errors coming out of `trigger_error()`.  While the current ErrorHandler does show a stack trace in the HTML output, the CLI output does not include stack trace information. Having to click into the stack trace for HTML messages, and having not hints at all in CLI contexts are some of the reasons I think this function could be useful to people upgrading.

markstory committed:
Add an alias for Deprecated to be compatible with PHP5.6

dereuromark commented on src/Core/functions.php:
```diff
@@ -254,10 +254,28 @@ function env($key, $default = null)
-    function deprecationWarning($message)
+    function deprecationWarning($message, $stackFrame = 2)
     {
+        if (!Configure::read('debug')) {
```
If we do this, then we wont be able to find any left overs in production.

markstory replied:
Yes, but I'm not comfortable adding potentially 100s of calls to `debug_backtrace()` to people's applications in production mode. Would it be better to check the error level here instead?

dereuromark replied:
Yeah, I think so.

inoas replied:
I think we should have a separate flag for this. Something along `log3xDeprecations` (default true if not set), which should go into a dedicated log file. That way it is easy to spot in production and easy to disable if you just want to run your app and not update it.

markstory committed:
Toggle deprecation warnings on error level.

By using the error level to control deprecation warnings, we allow users
to have them enabled in production but go into log files. This is useful
for ensuring all the deprecation warnings have been caught/fixed.

ravage84 commented on src/Core/functions.php:
```diff
@@ -248,3 +248,34 @@ function env($key, $default = null)
+        if (!(error_reporting() & E_USER_DEPRECATED)) {
+            return;
+        }
+        $trace = debug_backtrace();
```
May be a white line above

ravage84 commented on tests/TestCase/Core/FunctionsTest.php:
```diff
@@ -42,4 +42,39 @@ public function testEnv()
     }
+
+    /**
+     * Test error messages coming out when debug is on
```
"Test error messages coming out when debug is on, manually setting the stack frame"

ravage84 commented on tests/TestCase/Core/FunctionsTest.php:
```diff
@@ -42,4 +42,39 @@ public function testEnv()
+    }
+
+    /**
+     * Test error messages coming out when debug is on
```
"Test error messages coming out when debug is on, not setting the stack frame manually"

ravage84 reviewed (commented):
Beautiful!

markstory committed:
Address feedback from review.

markstory committed:
Windows path fix.
